### DISCUSSION TITLE
자료구조 - 배열/연결리스트 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#37] [질의 응답]
- 배열과 연결리스트의 특징에 대해서 설명해주세요.
- Static Array와 Dynamic Array의 차이점은 무엇일까요?
- Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?
- 연결리스트의 종류에 대해 설명해주세요.
- 배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.
- 연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.

**Comment by seongeun42** [comment]
### ❓️배열과 연결리스트의 특징에 대해서 설명해주세요.

<br>

### 💬 답변
<blockquote>

* **배열**
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/4f3bb1f5-e8ce-47dd-a2f1-eae66daa854f)
  * 데이터가 메모리 상에서 연속적으로 존재함
  * index로 접근 가능
  * index로 조회할 때 시간복잡도가 `O(1)`
    * 연속적이기 때문에 `배열의 첫번째 주소 + index` 만큼 하면 바로 조회 가능하기 때문
  * 공간이 연속적이기 때문에 삽입과 삭제가 쉽지 않음
    * 삽입할 때는 추가 공간을 확보해야 하고, 삭제할 때는 중간을 비워둬선 안 되기 때문에 뒤의 요소를 전부 하나씩 앞당겨야 함

<br/>

* **연결리스트**
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/a23f97b4-7a34-4c59-8757-c7560835505e)
  * 논리적으로는 연속적이나, 메모리 상에서는 연속적으로 존재하지 않음
  * 연결리스트를 이루는 노드가 유형에 따라 앞, 뒤 노드의 주소값을 가지고 있기 때문에 논리적으로 연속된 것처럼 보일 수 있음
  * 조회할 때 시간복잡도 `O(n)`
    * 특정 노드를 찾기 위해 첫번째 요소부터 특정 노드까지 찾아가야 하기 때문
  * 삽입/삭제할 때 시간복잡도 `O(1)`
    * 단, 삽입/삭제할 위치로 찾아가는 시간이 있기 때문에 실질적으로는 `O(n) + O(1)`

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️Static Array와 Dynamic Array의 차이점은 무엇일까요?

<br>

### 💬 답변
<blockquote>

- Static Array
  - 정적 배열
  - 크기가 고정되어 있음
  - 배열을 사용하기 전 미리 크기를 정해두고 사용함
  - 사용할 만큼의 크기를 정해두기 때문에 메모리 낭비가 없음

- Dynamic Array
  - 동적 배열
  - 크기가 가변적임
  - 크기를 정해두지 않고 요소를 추가하는 대로 늘릴 수 있음
  - 사용할 크기보다 더 큰 공간을 미리 할당받기 때문에 메모리 낭비가 있을 수 있음

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️Java에서 ArrayList는 어떤 식으로 크기를 늘리나요?

<br>

### 💬 답변
<blockquote>

따로 설정하지 않을 시 기본적인 크기는 10으로 시작한다. 이후 원소의 개수가 꽉 차면 `기존 크기 + 기존 크기 / 2`, 즉 1.5배만큼의 공간을 더 늘리고 기존 배열을 복사해 새로운 배열을 만들어내는 식으로 동적 배열을 구현한다. 이러한 방식 볼 때, 너무 자주 크기를 늘리면 자원이 낭비되므로 예상되는 크기가 크다면 초기 크기를 크게 설정해 배열을 만드는 것이 좋다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?

<br>

### 💬 답변
<blockquote>

Random Access란 Data Structure내의 다른 data에 의지하지 않고 index나 key값 등으로 순서 상관 없이 조회할 수 있는 접근법을 의미한다.
이와 반대되는 Access 개념으로는 Sequential Access가 있다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️Sequential Access의 예시를 하나 이상 들어주세요.

<br>

### 💬 답변
<blockquote>

연결리스트가 대표적인 Sequential Access이다. 여러 개의 노드로 이루어진 연결리스트는 노드에 저장된 다음 혹은 이전 노드 주소를 읽어 찾아감으로써 N번째 노드를 찾기 위해 첫 노드부터 N번째 노드까지 순차적으로 접근해야 한다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 연결리스트의 종류에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

- **단순 연결리스트**
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/4a262e7a-a908-4cc8-ac2f-7474b29ec331)
  - 첫 노드에서 마지막 노드까지 한 방향으로 연결되어 있는 연결리스트
  - 각 노드는 자신의 다음 노드의 위치 값을 가짐

<br>

- **이중 연결리스트**
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/6bbcb964-2847-4cf1-a23e-2d95bb7aaf7c)
  - 각 노드가 자신의 앞, 뒤 노드의 위치 값을 가짐

<br>

- **원형 연결리스트**
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/19a05336-9853-4e68-84c7-ad596abd59e1)
  - 마지막 노드가 첫번째 노드의 위치 값을 가짐


</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

배열은 메모리 상에 연속적으로 저장되어 있어 조회 속도가 O(1)로 빠른 한편 데이터의 삽입, 삭제는 느리다는는 특징이 있으므로, 데이터의 삽입, 삭제보다는 조회가 많이 일어나는 상황에 사용하는 것이 유리하다.

연결리스트는 논리적으로는 연속적이나 메모리 상에 비연속적으로 저장되어 있기 때문에 삽입, 삭제가 용이하나 조회 속도가 O(n)으로 느리다는 특징이 있으므로 데이터의 조회보다는 삽입, 삭제가 빈번한 상황에 사용하는 것이 유리하다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.

<br>

### 💬 답변
<blockquote>

연결리스트는 값과 다음 노드의 주소를 가지는 노드들로 이루어져 있다.
연결리스트에 요소를 추가할 때는 가장 마지막 노드가 가지는 다음 노드를 가리키는 변수에 추가하고자 하는 노드의 주소를 넣으면 된다.
반대로 요소를 삭제하고자 할 때는 삭제하고자 하는 요소 앞 노드가 가진 주소를 삭제할 요소가 가진 다음 요소의 주소로 바꾸면 된다.

</blockquote>

<br>
